src/http-gateways/path-gateway.md

```diff
@@ -567,7 +591,10 @@ The following response types require an explicit opt-in, can only be requested w
 - Raw Block (`?format=raw`)
   - Opaque bytes, see [application/vnd.ipld.raw](https://www.iana.org/assignments/media-types/application/vnd.ipld.raw).
 - CAR (`?format=car`)
-  - Arbitrary DAG as a verifiable CAR file or a stream, see [application/vnd.ipld.car](https://www.iana.org/assignments/media-types/application/vnd.ipld.car).
+  - A CAR file or a stream that contains all blocks required to trustless verify the expressed query, see [application/vnd.ipld.car](https://www.iana.org/assignments/media-types/application/vnd.ipld.car).
+    - Must contain, as the very first block, the block that hashes to the content root CID
+    - Must contain, immediately following the root block, all blocks encountered while traversing the expressed path in the order they were traversed
+    - Must contain, immediately following traversed path blocks, appropriate blocks in depth first traversal order required to verify the query expressed at the terminus of the path in [query parameters](#query-parameters-for-car-requests)
```

Comment:
I think we should clarify here whether it is expected that responses here should be byte-wise identical. I'd like us to explicitly make this a yes/no rather than having people guess.

Some ways in which you could have a CAR file not contain exactly the same bytes despite having a consistent traversal order:
- Different CAR version (e.g. CARv1 vs CARv2)
- Adding extra CAR features (e.g. indices, or if we add a solution for signaling EOF in CARs as a remedy for https://github.com/ipfs/specs/pull/332 while still calling it a CAR, etc.)
- How many and which CIDs are in the header

My inclination is to explicitly call out that `format=car` **does not reproduce deterministic results**. If people want reproducible results (e.g. for whatever DAG House is doing with CAR CIDs, or what Filecoin is doing with converting CARs to Piece CIDs) then we should add more parameters that describe exactly what it is they want so that we can change the defaults later.

This is similar to the situation we're in with `ipfs add` in kubo.

Reply:
Are you suggesting that block ordering and block inclusion could be non-deterministic too? If we could lock that in then it satisfies a whole lot of concerns around incremental verifiability.

But on the other ways that a CAR could be non-deterministic—I think the main problem could be the possibility of abuse in being able to craft CARs that are excessively large and wasteful. CARv2 has optional padding on both sides of the payload, so you can make them very large. A malicious server could begin a CARv2 with a very large data padding and just drip-feed feed random bytes. Or an index padding that's excessively long—which can be protected against on the client side (when CARv2, stop reading after CARv1 payload), but it's yet more logic to be inserted into this process just to protect against a class of misbehaviour.

Reply:
I'm considering turning this doc into an extension to the CARv1 spec for "deterministic CARs", similar to how some data format specs, such as CBOR, have sections with recommendations for deterministic forms: https://github.com/filecoin-project/lassie/blob/main/docs/CAR.md, it could serve as an explication of the 3 dot-points listed here.

Reply:
(and I'd like to get a good set of fixtures for these too, which we have partly in our Lassie test suite, but they need to be more portable and cover a few more cases)

Reply:
Pulling in @John-LittleBearLabs's comment above re responses _"containing as large a subset of this that can be fetched immediately"_: I think we could make that allowable as long as they are still in the right traversal order. What I want to be able to do is traverse the graph on the client side of this to confirm that the blocks are as expected. I _could_ error when the response contains a surprise block that's not the one I expected, alternatively I could skip over the ones I expect, exhausting the traversal, assuming that the block I was given might be later in the traversal because the server had dead-ends that it couldn't serve. It would be a shame to bail at the first sign of a missing block if the server has a lot more of the graph available.

What I don't want to do is have a memory pool of random blocks I was given that _may or may not_ be related to the graph I care about. I could do this with a single block while I run through a traversal looking for where the gap in blocks ends and where to pick up again.

That nicely leaves open the option to go fishing elsewhere for other parts of the graph that are missing, but maximising what a single server can provide.

(In many ways we're defining a dumber graphsync with this protocol so can learn lessons from it: https://github.com/ipfs/go-graphsync/blob/1988c8f1920ba9e558dd85ceaf84c0c76e0f46de/graphsync.go#L129)

Reply:
To push back on myself slightly, it might be worth considering the interplay weirdness between allowing gaps to exist (partial response) and a byte range (`bytes=x:y`). To be on the safe side, perhaps it should be considered an invalid request if you try to do both at the same time.

Reply:
Well, this is where client validation comes in - it's still an erroneous response as far as a client is concerned. Any "verifiable CAR" validation tools we write should say that "this is not a verifiable CAR". But it is a useful bundle of blocks we can use to work toward a verifiable CAR.

Until we have better error signalling mechanisms then we're a bit stuck with either "terminate" or "keep going and try our best" on the server side.

Reply:
> Are you suggesting that block ordering and block inclusion could be non-deterministic too? If we could lock that in then it satisfies a whole lot of concerns around incremental verifiability.

My main ask here is to define very explicitly if responses should be bytewise identical, and then if not what constraints and freedoms are implementers allowed. There are some people in our ecosystem who have been encoding CAR files in a very specific way and then assuming/hoping that other tooling behaves the same way or continues to behave exactly as it does today. This leads to similar problems as with people assuming kubo's `ipfs add` will always produce the same UnixFS graph by default.

I explicitly gave a few examples that had nothing to do with incremental verifiability or anything IPLD traversal related to demonstrate the potential flexibility here.

However, yes even in the realm of incremental verifiability there are multiple plausible ways to do block ordering and inclusion. For example:
- Instead of doing a specific kind of traversal we could have any valid traversal traversal order (e.g. for a UnixFS HAMT enumeration the blocks could come in any order rather than just a depth first one)
- Duplicate blocks could or could not be included. Duplicate blocks trade off more (although gzip might help here) data transfer in exchange for clients needing less memory if they're streaming outputs out. It's not unfathomable that both are desirable options for different use cases

> I'm considering turning this doc into an extension to the CARv1 spec for "deterministic CARs"

Sure, we can nail down both a spec for a deterministic CARv1 and a specific deterministic response type here (they might be the same or come with extra conditions around say what goes in the headers) if people would like this to be deterministic.

While I don't feel strongly about names inevitably someone will come around with a different way of creating the CAR (they want CARv2s, they want CARv2s with different indices, they want/don't want duplicate blocks, etc.) and will need a way to signal that. I don't know if `format=car` should mean "one of the many ways to encode this graph into a CAR that was blessed in mid 2023" or if perhaps we want a more explicit signal indicating that the format requested is the precise deterministic format being specified rather than just a valid CAR.

Reply:
> call out that format=car does not reproduce deterministic results

I've removed mentions of ordering/determinism from this IPIP and added call out (https://github.com/ipfs/specs/pull/402/commits/98191b5fec841d6ed67975efc457744373901b66).
I think we want to move this to a separate IPIP, as noted in https://github.com/ipfs/specs/pull/402#issuecomment-1544819745. 

My current thinking is (notes for future IPIP):
-  this type of thing should be explicit (a flag for order, and one for duplicates), not implicit, 
- clients need a mechanism to explicitly opt-in into specific CAR flavor (`Accept: application/vnd.ipld.car;ord=dfs;dup=t;q=0.9, application/vnd.ipld.car`  header or query params, to avoid issues with HTTP caching, `Etag`, and legacy gateways), 
- gateways / implementers shoudl be free to decide what is the implicit default for this, but then be clear about it in very explicit `Content-Type: application/vnd.ipld.car;version=1;ord=dfs;dup=t` present in responses

Reply:
Responding to https://github.com/ipfs/specs/pull/402#issuecomment-1544453059

👍 on not restricting `format=car` to have a defined byte output, I just don't think we're at a place where we can do that yet given the lack of consensus on the "best" way to return a CAR.

> Ultimately we're talking about balancing work between server & client. Less constraints makes it easier to write a performant server, and harder to write a performant client. More constraints does the opposite.

👍

> I'd rather avoid a complex signaling mechanism

Idk how complicated this is going to end up being, my understanding of the existing proposals are to:
1. Just ask the server what they support and then choose what you want explicitly (or yolo and don't specify if you don't care)
2. Maybe allow passing options using standard HTTP negotiation around Accept headers, etc.

IIUC these are both pretty normal HTTP negotiation tactics and negotiation here makes sense since "we're talking about balancing work between server & client" is roughly equal to "the client and server to need to **negotiate** how much work each of them will do".

> around traversal order on first pass and just say depth first. We can building an ordering parameter later.

If we were going to indicate that the blocks MUST be in a traversal order (as opposed to say having the root block at the end of the CAR) AND that we should specify that traversal order then DFS seems reasonable for most cases of UnixFS data. However, literally any traversal order would solve the "I downloaded a large amount of garbage before I realized I didn't want this data" (otherwise known as incremental verifiability) problem. Even not defining the traversal order (as long as it's a valid traversal) would work, albeit with perhaps more memory usage.

From a Slack thread:
> We cannot ... put whole CAR files (likely many at once) in memory or on disk before they verify a single block

IIUC (someone please correct me if I'm wrong) there are currently no people advocating for the data to come in a non-traversal order (e.g. the natural UnixFS encoding order where the root is at the end of the CAR). As long as some traversal order is used this problem goes away.

> Holding the whole response, either in RAM or disk, in order to verify it feels like a deal breaker for most clients. There's a secondary concern around time to verification failure

Does this indicate that you'd be in favor of having duplicate blocks supported in some way? Not having the ability to support duplicate blocks in the current CAR format means that you need to keep the entire response in RAM or disk to verify the response unless you are willing to use caching heuristics and follow-up your CAR requests with block requests to fill in the blanks for any deduplicated blocks that you deleted despite needing them.

IIUC the tradeoff of yes/no on duplicate blocks is that not having duplicate blocks is:
- Good: Fewer bytes over the wire
- Ok: Works reasonably if the data is being written to disk
- Bad: In order to implement streaming responses the client logic has to be more complex and potentially make additional network requests to compensate for not keeping all the data buffered in RAM/disk

Reply:
I've tried to synthesize  things, and extracted CAR block ordering spec work to [IPIP-412](https://github.com/ipfs/specs/pull/412), let's continue ordering discussion there.